[Round2] Step2 - hansol

enqnsol614 opened the pull request:
@jaemuYeo
[Round2] Step2 PR요청 드립니다.

## 고민했던 점
 ### 주제
- 이번 과제에서는 로또 번호 생성기를 구현해보았습니다. 이를 통해 1부터 45까지의 숫자 중 중복 없이 6개의 숫자를 무작위로 선택하는 로직을 구현하려고 했습니다. if lottoNumber > 0 && lottoNumber <=45 {} 이런식으로 해야할지 알았는데, 랜덤 번호를 구할 수 있는 매서드를 알게 되어 활용 했습니다. 그래서 if 문에는 중복인지 아닌지만 체크했습니다. 
 ---

 ## 해결이 되지 않은 점
 ### 주제
 -
 ---

 ## 조언을 얻고 싶은 부분
 ### 주제
 - getLottoNumber() 랜덤 숫자를 구하고, generateWinningNumbers()를 통해 배열 길이를 조절하고 최종적인 로또번호생성기 느낌이었는데, 하나의 함수로 만드는게 나았는지 모르겠어요. 함수를 나누고, 함수명 짓는게 적절한건지 조언을 얻고 싶습니다.
 - checkMatchingNumbers(myLottoRound: myLottoRound, myLottoNumbers: myLottoNumbers, currentLottoRound: lottoGenerator.currentLottoRound, winningNumbers: lottoGenerator.winningNumbers, bonusNumber: lottoGenerator.bonusNumber) 이렇게 매개변수가 길어지는게 맞나요,,?

Commit:
add LottoNumberGenerator.swift file.

Commit:
Add functionality for matching my numbers, current lotto round info, and creating bonus number

Commit:
check current lotto round

Commit:
Change round information to random number

Commit:
Add print message for when only the bonus number is matched

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,63 @@
+
+import Foundation
+
+class LottoNumberGenerator {
```
오우!! 이번 Round에서는 객체를 다루지 않습니다 🙏
너무 좋은 시도였지만 요구사항에 맞게 함수만으로 구현되면 좋을 것 같아요 :)

enqnsol614 replied:
줌 강의를 듣고 활용해보았습니다ㅎㅎ
요구사항에 없으므로 해당 내용 삭제하여 커밋 했습니다!

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,63 @@
+    var bonusNumber: Int = 0
+    var currentLottoRound: Int = Int.random(in: 1...2)
+    
+    func getLottoNumbers() {
```
numbers 처럼 복수형으로 네이밍 된 것이 너무 좋네요! 👍

enqnsol614 replied:
감사합니다!

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,63 @@
+            winningNumbers.append(lottoNumber)
+        }
+        
+        repeat {
```
여러 반복구문 중 repeat-while문을 사용한 이유가 궁금해요! 🤔

enqnsol614 replied:
repeat 블록 안에는 최초 1회 선 실행할 코드이면서 조건에 맞을때까지 반복되어야하는 코드가 들어가고
while 블록 안에는 조건문을 써야하는 것이므로
보너스 번호는 최초 1회 무조건 생성되어야 하고, 중복이 되면 다시 생성하도록 하면 되기 때문에 해당 반복문이 적합하다고 생각했습니다.

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,63 @@
+    }
+    
+    func generateLottoWinningNumbers() {
+        winningNumbers = []
```
winningNumbers를 빈배열로 초기화 시킨 이유는 무엇인가요?

enqnsol614 replied:
for _ in .. 반복문을 이용하여 함수 printLottoWinningNumbers 를 호출할 경우  배열이 초기화되지 않는 오류가 있어서
추후에 반복하여 호출할 경우를 방지하고자 winningNumbers = [] 를 초기화 해주었습니다. 사실 1번 호출할 땐 필요없는 게 맞는데 이럴 경우엔 그냥 삭제하는게 좋은 가요?!

반복하여 함수 호출:
for _ in 1...3 {
    printLottoWinningNumbers()
}

초기화 하지 않을 떄 결과값:
1회차의 당첨 번호는 [3, 11, 20, 21, 23, 28], 보너스 번호는 36입니다.
1회차의 당첨 번호는 [3, 11, 20, 21, 23, 28], 보너스 번호는 36입니다.
1회차의 당첨 번호는 [3, 11, 20, 21, 23, 28], 보너스 번호는 36입니다.

초기화 했을 때 결과값 : 
1회차의 당첨 번호는 [3, 8, 15, 27, 28, 35], 보너스 번호는 20입니다.
1회차의 당첨 번호는 [16, 23, 24, 27, 28, 33], 보너스 번호는 5입니다.
1회차의 당첨 번호는 [2, 18, 20, 26, 31, 42], 보너스 번호는 23입니다.

jaemuYeo replied:
아뇨! 단순히 이유가 궁금했습니다 ㅎㅎ 좋아요! 👍

jaemuYeo reviewed (changes requested):
hansol님! 제출해주신 스텝이 Step2 맞으실까요?
실행예시가 스텝3과 비슷해서 리뷰 중 잠시 멈추고 확인 후에 다시 진행하도록 하겠습니다!

enqnsol614 commented:
@jaemuYeo 
Step3 내용을 보지 않고 개인적으로 로또사이트 보면서 시도해보고 싶은 것들을 시도해 본거 였는데.. 저도 하고 나서 다음 강의들을 듣고 보니 스텝3에 '회차'에 대한 내용이 있더라구요ㅜㅜ 다시 하는게 좋을까요??

Commit:
Delete the class object.

jaemuYeo commented:
순차적으로 진행해주시면 좋을 것 같아요 :)
요구사항에 맞게 진행하여 동일한 결과를 불러오는 것도 연습이라고 생각해주시면 좋을 것 같습니다!!
조금만 수정하시면 금방 하실 것 같아요~ 백업해두셔도 좋습니다! 💪
객체사용은 라운드 3에서 있으나 이미 구현하셨으니 두셔도 좋아요!

Commit:
Delete currentLottoRound info.

enqnsol614 commented:
@jaemuYeo
넵 ! 회차에 관한 내용 삭제 하였습니다!!

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,55 @@
+    } while winningNumbers.contains(bonusNumber)
+}
+
+func generateLottoWinningNumbers() {
```
저는 난수를 생성하여 append하는 부분과 이 부분을 분리해놓은 것이 인상 깊어요 👍
좋은 시도였습니다!

enqnsol614 replied:
@jaemuYeo 
감사합니다!!

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,55 @@
+
+func printLottoWinningNumbers() {
+    generateLottoWinningNumbers()
+    print("이번 회차의 당첨 번호는 \(winningNumbers), 보너스 번호는 \(bonusNumber)입니다.")
```
이 부분과 49번 라인의 print결과물이 달라요!
숫자들이 배열에 감싸져있는 것을 확인할 수 있는데, 프로젝트의 실행예시를 한 번 더 관찰해보시기 바랍니다 :)
[참고 블로그](https://sunny-maneg.tistory.com/entry/swift-Array-%EB%A9%A4%EB%B2%84-%EB%8C%80%EA%B4%84%ED%98%B8%EC%97%86%EC%9D%B4-%ED%95%9C-%EC%A4%84%EB%A1%9C-%EC%B6%9C%EB%A0%A5%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95-joind-map) 를 통해 해결해보시고 꼭 공식문서에서도 관련 내용을 찾아보시면 좋을 것 같아요 💪

enqnsol614 replied:
배열이 그대로 들어가도록 코딩이 되어있었군요!
링크해주신 블로그를 참고하여 joined 메서드를 사용하도록 수정했습니다.

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,55 @@
+    print("이번 회차의 당첨 번호는 \(winningNumbers), 보너스 번호는 \(bonusNumber)입니다.")
+}
+
+func checkMatchingNumbers(myLottoNumbers: [Int], winningNumbers: [Int], bonusNumber: Int) {
```
매개 변수가 많고, 함수 길이가 긴 것을 실제 애플이 만들어놓은 프레임워크내에서도 존재합니다!
다만 현재 함수에서는 조금 수정이 필요한 부분이 있을 것 같아요.

1. winningNumbers와 bonusNumber를 내부에서 구현하지않고 외부에서 주입받아야하는 이유가 있나요~?
2. check와 matching 중 하나는 제거해도 될 것 같아요. 간결하게 matchNumber로만 해주어도 충분할 것 같습니다!

enqnsol614 replied:
1. 불필요한 매개변수 삭제하였습니다!
2. 함수명 변경하였습니다

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,55 @@
+    } else {
+        print("축하합니다! 겹치는 번호는 \(matchingNumbers) 입니다!")
+        
+        if bonusMatch {
```
들여쓰기가 한 번 더 된것 같아요!

enqnsol614 replied:
어느 부분이 들여쓰기가 된 것 인지 못찾았습니다!ㅠㅠ
else 구문안에 if문이 또 들어간거 얘기하시는 걸까요???

jaemuYeo replied:
앗 이건 제가 잘 못봤군요! else문 외부에 있는줄 알았어요 🙏

jaemuYeo reviewed (changes requested):
빠르게 수정해주신줄 모르고 확인이 조금 늦었네요 🙏
몇 가지 코멘트 남겨두었습니다! 해결 후 Merge 진행할게요 :)

Commit:
1. Updated variable names.
2. Modified 'matchingNumbersString' by converting array elements into a joined string

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,59 @@
+
+func printLottoWinningNumbers() {
+    generateLottoWinningNumbers()
+    let sWinningNumbers = winningNumbers.map { String($0) }.joined(separator: ", ")
```
앞아 붙은 s는 무엇을 의미하나요?

jaemuYeo commented on CodeStarterCamp_Week2/LottoNumberGenerator.swift:
```diff
@@ -0,0 +1,59 @@
+    let matchingNumbers = myLottoNumbers.filter { winningNumbers.contains($0) }
+    let bonusMatch = myLottoNumbers.contains(bonusNumber)
+    
+    if matchingNumbers.isEmpty && !bonusMatch {
```
나중에 과정 종료 후 switch-case로 리팩토링 해보아도 좋을 것 같아요 :)
한 번쯤은 꼭 해보세요!

jaemuYeo reviewed (commented):
결과물과 코드 가독성 등 잘 이루어진 것 같습니다!!
몇 가지 남겨둔 코멘트 있으니 확인해주세요 :)
이번 스텝 Merge 진행하겠습니다!
다음 스텝은 이전에 했던 결과물이 있어서 금방하실 것 같아요 💪
